src/marks/text.js

```diff
@@ -377,21 +397,65 @@ const defaultWidthMap = {
 // https://exploringjs.com/impatient-js/ch_strings.html#atoms-of-text
 function defaultWidth(text, start, end) {
   let sum = 0;
-  for (let i = start; i < end; ++i) {
-    sum += defaultWidthMap[text[i]] || defaultWidthMap.e;
-    const first = text.charCodeAt(i);
-    if (first >= 0xd800 && first <= 0xdbff) {
-      // high surrogate
-      const second = text.charCodeAt(i + 1);
-      if (second >= 0xdc00 && second <= 0xdfff) {
-        // low surrogate
-        ++i; // surrogate pair
-      }
-    }
-  }
+  for (let i = start; i < end; i += glyphLength(text, i)) sum += defaultWidthMap[text[i]] || defaultWidthMap.e;
   return sum;
 }
 
 function monospaceWidth(text, start, end) {
-  return end - start;
+  let sum = 0;
+  for (let i = start; i < end; i += glyphLength(text, i)) sum += 100;
+  return sum;
+}
+
+function overflow(input, width, widthof, textOverflow) {
+  switch (textOverflow) {
+    case "clip-end":
+      return clip(input, width, 1, widthof, "");
+    case "ellipsis-end":
+      return clip(input, width, 1, widthof, "…");
+    case "clip-start":
+      return clip(input, width, 0, widthof, "");
+    case "ellipsis-start":
+      return clip(input, width, 0, widthof, "…");
+    case "ellipsis-middle":
+      return clip(input, width, 1 / 2, widthof, "…");
+  }
+}
+
+// Clips a text to the given width, balancing head and tail in proportion to p.
+function clip(text, width, p, widthof, insert) {
+  if (insert) width -= widthof(insert, 0, insert.length);
+  text = text.trim();
+  const head = [];
+  const tail = [];
+  const lengths = [];
+  let dropped = 0;
+  let w = 0; // width consumed by selected chars
+  for (let i = 0; i < text.length; ) {
+    const char = text.slice(i, (i += glyphLength(text, i)));
+    const l = widthof(char, 0, char.length);
+    if (w < width * p) {
+      head.push(char), (w += l);
+    } else if (w < width) {
+      tail.push(char), lengths.push(l), (w += l);
+      while (w >= width) tail.shift(), (w -= lengths.shift()), dropped++;
+    } else {
+      dropped++;
+      if (dropped > 1) break;
+    }
+  }
+  return dropped <= (insert ? 1 : 0) ? text : head.join("").trimEnd() + insert + tail.join("").trimStart();
+}
+
+// The size of the glyph at the given index. Supports the zero width joiner for
+// composed emojis. TODO: support more modifiers (e.g. fitz).
+function glyphLength(text, i) {
+  let a;
+  let j = 0;
+  do {
+    a = text.charCodeAt(i + j++);
```

Comment:
The mixed use of `j++` and `++j` is likely wrong here; doesn’t this mean we’re reading the same character we previously read on line 458 in the case of a surrogate pair?

Reply:
It's deliberate and I think, correct—this only happens when the codePoint is the zero width joiner. But I should probably set up unit tests.

Reply:
Hmm, I’ll investigate further as I’m not convinced. 🤔 It means that we’re reading the same character twice.

Also, smaller nit, but I think this function should return the index of the next character/glyph so that the calling code can say `i = readGlyph(text, i)` (slightly simpler) instead of `i += glyphLength(text, i)`.

Reply:
yes; I also wanted to explore making a generator that yields each glyph in turn, but it started to impact the line wrapping code.